src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,395 @@
+
+[mixed-mode programs]: /guides/language/evolution#mixed-mode-programs
+
+We recommend that you **first migrate leaf libraries** —
+libraries that don't import other files from the package.
+Then migrate libraries that directly depend on the leaf libraries.
+End by migrating the libraries that have the most
+intra-package dependencies.
+
+For example, say you have a `lib/src/util.dart` file
+that imports other (null-safe) packages and core libraries,
+but that doesn't have any `import '<local_path>'` directives.
+Consider migrating `util.dart` first,
+and then migrating simple files that depend only on `util.dart`.
+If any libraries have cyclic imports
+(for example, A imports B which imports C, and C imports A),
+consider migrating those libraries together.
+**_[PENDING: is everyone OK with those recommendations and examples?]_**
+
+To migrate a package by hand, follow these steps:
+
+1. Modify the SDK constraints in the package's `pubspec.yaml` file,
+   setting the language version to 2.12 beta or later:
+   ```yaml
+environment:
+  sdk: '>=2.12.0-18.0.beta <3.0.0'
+```
+
+2. Run `dart pub get`. <br>
+   ```terminal
+$ dart pub get
+```
+
+   Running `dart pub get` with a lower constraint of 2.12.0
+   sets the default language version of
+   every library in your package to 2.12,
+   opting them all in to null safety.
+
+
+3. Open the package in your IDE. <br>
+   You're likely to see a lot of analysis errors.
+   That's OK.
+
+4. _Optional:_ Add a [language version comment][] to the top of
+   any Dart files that you don't want to consider during your current migration:
+   ```dart
+// @dart=2.9
+```
+
+   Using language version 2.9 for a library that's in a 2.12 package
+   enables mixed mode and can reduce analysis errors (red squiggles).
+   However, **mixed mode reduces the
+   information the analyzer can use.**
+   For example, the analyzer might assume a
+   parameter type is non-nullable,
+   even though a 2.9 file might pass in a null value.
+
+5. Migrate the code of each Dart file,
+   using the analyzer to identify static errors. <br>
+   Add `?`, `!`, `required`, `late` and refactoring as needed
+   to improve your code and eliminate static errors.
+
+[language version comment]: /guides/language/evolution#per-library-language-version-selection
+
+**_[PENDING: Do we need to talk about `pub get` generating a [package config][]?
+That link isn't as helpful as it could be.
+Should/do we cover this on dart.dev?]_**
+
+[package config]: https://pub.dev/packages/package_config
+
+
+## Test your code
+
+If you already have tests, run them.
+You might need to update tests that expect null values.
+
+**_[PENDING:
+See [Experimenting with null safety in Flutter](https://github.com/flutter/flutter/wiki/Experimenting-with-null-safety-in-Flutter)
+for details (on testing and otherwise; it's sort of this doc, but for pre-beta.
+]_**
+
+
+## Publish your code
+
+We encourage you to publish non-app packages as prereleases
```

Comment:
This wording...worries me.

`publish non-app packages` – are we worried that folks confuse publishing an app (to a store) vs publishing a package (to pub.dev)?

maybe we should have seperate sections: `Publishing packages to pub.dev` vs `Publishing your application to a store`?

Reply:
I originally had two sections, but since we don't really want people to publish apps yet, we deleted that section. We don't really have anything to say about the app case, anyway.

I don't think most people think of apps as packages, so I should be able to reword without lying.

Reply:
I don't think we need to be so strong as to say "you cannot publish apps". We have plenty of Flutter beta features that are used in published apps.

I think we should make it clear that a beta feature is potentially less stable, and then let app developers choose themselves if they want to take the risk.

Reply:
We don't say "cannot publish apps". We just say "we generally recommend waiting for a stable SDK release."

Reply:
Well, we say "we generally recommend waiting for a stable SDK release". That seems a bit late to say here under step 4. We already have a general warning at the top of the page, so I think we can remove the warning here from step 4.